### PR TITLE
Publish metrics to MC and JMX

### PR DESCRIPTION
**Pull request by blazember**
Metrics are recorded to a ringbuffer that Management Center can read out
by using `ReadMetricsMessageTask`. The task reads the metrics recorded
since the last seen sequence number. The next sequence number that MC
can continue reading the Metrics with is returned in the response. The
ringbuffer, the `ReadMetricsOperation` and the task are almost identical
copies of the similar Jet classes.

Besides the described mechanism used for recording and publishing,
configuring the metrics system has also changed. Before this commit,
metrics could be configured with JVM arguments, now it can be done with
`MetricsConfig` both on the member and on the client side. This
configuration is shared with the diagnostic's `MetricsPlugin`.

The client protocol changes required for the `ReadMetricsMessageTask` is
also included in this commit.

### CODE REVIEW

**Comment by cangencer** [comment]
@blazember we've made since some additions to the metrics, maybe you want to have a look at the latest state? https://github.com/hazelcast/hazelcast-jet/tree/master/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics In general, the compressor and so on have been made into separate classes. We'd also need some hooks to register additional publishers. (see JobMetricsPublisher) for example.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+                    1, (int) Math.ceil((double) config.getRetentionSeconds() / config.getCollectionIntervalSeconds())
+            );
+            metricsJournal = new ConcurrentArrayRingbuffer<>(journalSize);
+            ManagementCenterPublisher publisher = new ManagementCenterPublisher(this.nodeEngine.getLoggingService(),
+                    (blob, ts) -> {
+                        metricsJournal.add(entry(ts, blob));
+                        pendingReads.forEach(this::tryCompleteRead);
+                    }
+            );
+            publishers.add(publisher);
```
Comment (eminn) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: Is there any mechanism where we can register additional metric publishers ?

File: hazelcast/src/main/java/com/hazelcast/internal/util/BiTuple.java [hunk 2]
```diff
@@ -24,18 +24,18 @@
  * @param <Y> type of the second element
  */
 @SuppressWarnings("checkstyle:visibilitymodifier")
-public final class Tuple2<X, Y> {
+public final class BiTuple<X, Y> {
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/internal/util/BiTuple.java]: wouldn't it be better to send this another PR, also including Tuple3 (which also exists in Jet?)

Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/util/BiTuple.java]: Yes, it would be better, given the number of places affected. I send it in a separate PR and changing this to draft as long as the tuple PR is not merged.

File: hazelcast/src/main/java/com/hazelcast/config/MetricsConfig.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+     */
+    public int getCollectionIntervalSeconds() {
+        return this.intervalSeconds;
+    }
+
+    /**
+     * Sets whether statistics for data structures are added to metrics.
+     * It's disabled by default.
+     * <p/>
+     * Note that enabling the data structures metrics also set {@link #minimumLevel}
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/config/MetricsConfig.java]: A potential typo: `also set` should be `also sets`.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+     * function that takes {@link NodeEngine} as input letting the caller to
+     * optionally initialize the publisher returned from the function.
+     *
+     * @param registerFunction The function that returns with the {@link MetricsPublisher}
+     *                         instance.
+     */
+    public void registerPublisher(Function<NodeEngine, MetricsPublisher> registerFunction) {
+        MetricsPublisher publisher = registerFunction.apply(nodeEngine);
+        publishers.add(publisher);
+    }
```
Comment (blazember) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: @cangencer @eminn :point_up:

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+            ManagementCenterPublisher publisher = new ManagementCenterPublisher(this.nodeEngine.getLoggingService(),
+                    (blob, ts) -> {
+                        metricsJournal.add(entry(ts, blob));
+                        pendingReads.forEach(this::tryCompleteRead);
+                    }
+            );
+            publishers.add(publisher);
+        }
+
+        if (config.isJmxEnabled()) {
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: we also changed the logic here, so the `isEnabled` is like a master switch, I don't know how you want to make the configuration. so if isEnabled is false, you can't enable just JMX.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java (hunk 4 above)
Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: I think this will be fine, we can register the publisher in JetService.init()

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java (hunk 5 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: That means you can't disable publishing to MC (buffers) if you want JMX only. How if we have this in the config?
```
public boolean isMetricsEnabled() {
  return isMcEnabled() || isJmxEnabled();
}
```

And we could still switch MC or JMX independently.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/jmx/JmxPublisher.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            unregisterMBeanIgnoreError(metricData.objectName);
+        }
+        mBean.setMetricValue(metricData.metric, metricData.unit, value);
+    }
+
+    private void unregisterMBeanIgnoreError(ObjectName objectName) {
+        try {
+            platformMBeanServer.unregisterMBean(objectName);
+        } catch (InstanceNotFoundException ignored) {
+            // Unregistration can by racy, we unregister from multiple places.
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/jmx/JmxPublisher.java]: A typo: `by racy` should be `be racy`

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/ConcurrentArrayRingbuffer.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        if (startPoint >= endPoint) {
+            arraycopy(ringItems, startPoint, result, 0, capacity - startPoint);
+            arraycopy(ringItems, 0, result, capacity - startPoint, endPoint);
+        } else {
+            arraycopy(ringItems, startPoint, result, 0, endPoint - startPoint);
+        }
+        return new RingbufferSlice<>(result, tail);
+    }
+
+    public synchronized int getCapacity() {
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/ConcurrentArrayRingbuffer.java]: Does this method have to be synchronized, considering that `capacity` is `final`?

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    public List<MetricsCollection> collections() {
+        return collections;
+    }
+
+    /**
+     * Deserializing iterator for reading metrics
+     */
+    public static final class MetricsCollection implements Iterable<Metric> {
+
+        private final long timestamp;
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java]: Shouldn't this be `sequence` instead of `timestamp` for the sake of consistency?

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsCompressor.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            // should never be thrown
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void addDouble(String name, double value) {
+        try {
+            writeName(name);
+            // convert to long with specified precision
+            dos.writeLong(Math.round(value * DOUBLE_TO_LONG));
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsCompressor.java]: A really minor consideration. Looks like we're loosing information about the type of metric here. It's fine, but it creates additional coupling: MC has to know which metrics have `double` type. Maybe, it could be avoided by adding a special tag for such metrics (say, `type=double`) or an additional byte header with the type id. Then `decompressingIterator` could be restoring those `double`s (with some precision loss) and saving them into `Metric` objects.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java (hunk 5 above)
Comment (eminn) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: How about adding the enabled/disabled information accessible via MetricsService ? This way we don't need to go and fetch the Config object from the NodeEngine and access MetricsConfig from there. We already have MetricsService and it has the MetricsConfig.

Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: 2 cents from me. DS stats are enabled with the following check:
https://github.com/hazelcast/hazelcast/pull/15560/files#diff-b3e262104ca2692463bc0f125cb5fd22R217

If I'm getting everything right, it means that users won't be able to enable JMX and DS stats only, without enabling MC metrics. If it's correct, then it looks a bit weird to me.

Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: Yes, that's indeed better than doing it in the config.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java (hunk 8 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java]: The ringbuffer is typed with `Map.Entry<Long, byte[]>` where the key is a timestamp taken in `ManagementCenterPublisher#whenComplete`. That's where the timestamp is from.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java (hunk 5 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: @puzpuzpuz this is a valid point, especially if we consider registering custom publishers. I think 
- enabling DS metrics should be independent from enabling MC or JMX publishers
- enabling MC and JMX publishers should also be independent from each other

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java (hunk 8 above)
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java]: Yes, I know that. It's just strange that the same thing (timestamp) is called `sequence` in other APIs, like `HazelcastClientInstanceImpl#readMetricsAsync`'s arguments or `MetricsResultSet#nextSequence`.

Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java]: They're not the same thing, sequence is the sequence of the metrics in the ringbuffer, timestamp is the system time when it's collected.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsCompressor.java (hunk 9 above)
Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsCompressor.java]: It only supports `long` type of metric, and the only double metrics are things like cpu usage where you don't care about precision to that degree.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java (hunk 5 above)
Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: Well MC collection doesn't do anything by itself except to store the latest blob. If metrics will be rendered anyway I think you can always enable MC collection?

Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: > How about adding an access to enabled/disabled information via MetricsService ? This way we don't need to go and fetch the Config object from the NodeEngine and access MetricsConfig from there. We already have MetricsService and it has the MetricsConfig.

Well you pass a function to register publisher and if metrics are disabled, the function shouldn't be called?

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+    /**
+     * Register a custom {@link MetricsPublisher} implementation with a register
+     * function that takes {@link NodeEngine} as input letting the caller to
+     * optionally initialize the publisher returned from the function.
+     *
+     * @param registerFunction The function that returns with the {@link MetricsPublisher}
+     *                         instance.
+     */
+    public void registerPublisher(Function<NodeEngine, MetricsPublisher> registerFunction) {
+        MetricsPublisher publisher = registerFunction.apply(nodeEngine);
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: this function should only be called if metrics are enabled.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java (hunk 8 above)
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java]: > They're not the same thing, sequence is the sequence of the metrics in the ringbuffer, timestamp is the system time when it's collected.

They're tightly coupled at the moment. As an example, `startSequence` argument in `HazelcastClientInstanceImpl#readMetricsAsync` really means `startTimestamp`/`fromTimestamp` in terms of internal processing and it affects what is returned in `MetricsCollection`s, including `timestamp` field values.

I'm totally OK if we leave as is, but I'd prefer to use term `timestamp` in all "external" APIs to avoid confusion.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsCompressor.java (hunk 9 above)
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsCompressor.java]: It's not about precision loss, it's about lost value type. It's absolutely fine that the compression is lossy. But it would be nice if the iterator could recover types when data points are uncompressed. In this case, it'll be easier to support additional data point types in the future (see my considerations above).

And as I mentioned above, MC will work fine with those lost types, as it's aware of metrics and can convert metrics with double values based on their names.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java (hunk 5 above)
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: > If metrics will be rendered anyway I think you can always enable MC collection?

In this case, rendering and storing those blobs will consume some CPU time and memory even if blobs aren't used at all. I'd prefer to pay for it only when MC metrics collection is enabled.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java (hunk 8 above)
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java]: A follow-up. I've double checked the `ConcurrentArrayRingbuffer` implementation and now I see that `sequence` is position in the ring buffer, as you said. Thanks for the clarification @cangencer!

File: hazelcast/src/main/java/com/hazelcast/config/MetricsConfig.java (hunk 3 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/config/MetricsConfig.java]: Fixed

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java (hunk 5 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: Finally, introduced a master switch along with `mc-enabled` flag, so populating metrics to the ringbuffer can be turned off. The master switch is needed because we support programmatic publisher registration that is expected to ignore the registration if the metrics system is entirely switched off. It is a valid case that both MC and JMX publishers are disabled, but we want to allow programmatic registration of publishers. It can't be done without a master switch.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/jmx/JmxPublisher.java (hunk 6 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/jmx/JmxPublisher.java]: Fixed

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/ConcurrentArrayRingbuffer.java (hunk 7 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/ConcurrentArrayRingbuffer.java]: Feels a copy-paste leftover. Removed synchronized.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java (hunk 10 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java]: Done. We log on `fine` level if the metrics system is disabled.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsCompressor.java (hunk 9 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsCompressor.java]: I agree we need to carry the types. I'd address this is in a separate PR so that integration with MC side gets unblocked.

**Comment by blazember** [comment]
Things noticed in this PR and will be addressed in a separate PR:
- Add type information to the metrics for MC
- Client configuration should be reduced to `enable`, `jmx-enable` and `minimum-level`. The rest is relevant for member only.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+public class MetricsResultSet {
+
+    private final long nextSequence;
+    private final List<MetricsCollection> collections;
+
+    public MetricsResultSet(long nextSequence, List<Map.Entry<java.lang.Long, byte[]>> collections) {
+        //        this.nextSequence = slice.nextSequence();
+        this.nextSequence = nextSequence;
+        this.collections = collections.stream()
```
Comment (asimarslan) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java]: A minor improvement can be removing this convertion. 
We can just keep `List<Map.Entry<java.lang.Long, byte[]>>`  instead of a `List<MetricsCollection>`
MetricCollection just provide an iterator which delegates to `decompressingIterator(bytes)`
MC codebase can directly call this static method.

File: hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java [hunk 12]
```diff
@@ -168,6 +174,17 @@
 
     private static final AtomicInteger CLIENT_ID = new AtomicInteger();
 
+    private final ClientMessageDecoder<MetricsResultSet> decodeMetricsResponse = new ClientMessageDecoder<MetricsResultSet>() {
+
+        @Override
+        public MetricsResultSet decodeClientMessage(ClientMessage clientMessage) {
+
+            MetricsReadMetricsCodec.ResponseParameters response = serializationService
```
[the remark this answers is not in the document]
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java]: :+1:, removed

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java (hunk 11 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsResultSet.java]: Removed the conversion

**Comment by blazember** [comment]
Thank you a lot for the review rounds and for your help @asimarslan @cangencer @eminn @puzpuzpuz
